src/main/java/org/dave/compactmachines3/block/BlockRedstoneTunnel.java

```diff
@@ -81,16 +81,19 @@ public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, En
 
         if(faceX >= 0.69d && faceX <= 0.94d && faceY >= 0.69d && faceY <= 0.94d) {
             world.setBlockState(pos, state.withProperty(IS_OUTPUT, !state.getValue(IS_OUTPUT)));
-            WorldSavedDataMachines.INSTANCE.toggleRedstoneTunnelOutput(pos);
+            WorldSavedDataMachines.getInstance().toggleRedstoneTunnelOutput(pos);
 
             notifyOverworldNeighbor(pos);
             world.notifyNeighborsOfStateChange(pos, Blockss.redstoneTunnel, false);
         } else {
             EnumFacing nextDirection = StructureTools.getNextDirection(connectedSide);
 
-            int coords = StructureTools.getCoordsForPos(pos);
-            HashMap<EnumFacing, RedstoneTunnelData> sideMapping = WorldSavedDataMachines.INSTANCE.redstoneTunnels.get(coords);
-            while(sideMapping != null) {
+            int id = StructureTools.getIdForPos(pos);
+            Map<EnumFacing, RedstoneTunnelData> sideMapping = WorldSavedDataMachines.getInstance().redstoneTunnels.get(id);
+            if (sideMapping == null) {
+                return true;
+            }
+            while (true) {
```

Comment:
Single null check over a long-running while loop, looks like?

Reply:
This loop is weird. I didn't investigate it much, but IntelliJ notified me that the `sideMapping` can only be null once since it is not updated inside the loop. I'm not sure what the loop is doing, but I'd rather not change something I don't understand.